keps/sig-cli/20190318-kustomize-generators-transformers.md

```diff
@@ -0,0 +1,389 @@
+## Motivation
+
+1. Enable organizations to develop their own authoring solutions bespoke to their requirements, and to integrate
+   those solutions with kustomize and kubectl.
+1. Enable tools and authoring solutions developed by the ecosystem to natively integrate with kustomize and kubectl
+   commands.
+1. Enable Kustomize power users to augment Kustomize with their own *Transformers* and *Generators*.
+
+### Background
+
+Many users of Kubernetes require the ability to also write Config through high-level abstractions
+rather than only as low-level APIs.
+
+- Numerous tools and DSLs have been developed in the ecosystem in order to address this need
+  - Helm
+  - Ksonnet
+- Organizations have built their own tooling for generating Kubernetes Resource Config
+  - [kube-gen] (AirBnB) 
+
+Support for connecting the tools developed by the ecosystem with kubectl relies on piping commands together,
+however pipes are an imperative technique.  This KEP proposes a declarative approach for composing tools built
+in the Kubernetes ecosystem or bespoke tools developed by organizations that are tightly coupled to their
+environments.
+
+[kube-gen]: https://events.linuxfoundation.org/wp-content/uploads/2017/12/Services-for-All-How-To-Empower-Engineers-with-Kubernetes-Melanie-Cebula-Airbnb.pdf
+
+### Goals
+
+- Allow config authoring solutions developed in the ecosystem to be integrated into kubectl as declarative
+  Generators and Transformers
+- Allow users and organizations to develop their own custom configu authoring solutions and integrate
+  them into kubectl
+- Allow Kustomize power users to augment Kustomize's own Tranformers with their own
+- Allow users to perform client-side mutations so they they show up in diffs and are auditable
+
+### Non-Goals
+
+- Rewrite Kustomize on top of plugins
+  - This should be possible, but isn't a problem we need to solve right now
+- Develop a market place of any sort
+  - The ecosystem solutions themselves will have market places.  This allows those ecosystem tools to
+    plug their market places into kubectl via Kustomize.
+
+## Proposal
+
+Introduce 2 new types of `exec` plugins:
+
+- Transformers
+- Generators
+
+#### Terms
+
+- *Virtual Resource*
+  - Resource that is not a server-side Kubernetes API object, but used to configure tools
+  - Examples: kubeconfig, kustomization
+- *Generator*
+  - Kustomize directive that accepts a Virtual Resource as input, and emits non-Virtual Resources as output
+  - Examples: `configMapGenerator`, `secretGenerator`
+- *Transformer*
+  - Kustomize directive that accepts 1 Virtual Resource *and* collection of non-Virtual Resources as input
+    and emits non-Virtual Resources as output.
+  - Examples: `commonAnnotations`, `namespace`
+- *Built-In*
+  - Either a Transformer or Generator that is embedded directly in the `kustomization.yaml` rather than
+    as a separate API.
+
+#### Generators
+
+Generators generate 0 or more Resources for some virtual Resource input.
```

Comment:
This is something @monopole also suggested.

IMHO, it is simpler to provide 2 separate focussed and structured patterns through 2 APIs, rather than 1 API that can implement arbitrary patterns.  Mostly this is to simplify ordering and dependency considerations.

However I am open to being convinced someone can show me a solution that addresses dependency and ordering in a way that bounds the complexity users can introduce into the system.

Reply:
> show me a solution that addresses dependency and ordering

The unified version is just a single ordered list of "processors" (or whatever we call it) - the same as you already need for transformers.  The only hard bit is how you add in the current built-ins, and you already have this problem too (as raised elsewhere).

So the solutions are the same:
- You can just declare for each builtin whether it always gets executed first (builtin generators) or last (builtin transforms).
- Or alternatively, you can give them some representation in the explicit list and allow users to order the transforms explicitly (eg: `transforms: ["scale-resources.yaml", "builtin:nameSuffix"]`).

The only difference is that generator plugins are now listed in the transformers list too, instead of being in a separate unordered list.

I don't think it introduces any new complexity (and indeed removes some for both implementer and user) - they're just a transform that does "append" instead of pure generation.

Reply:
Appreciated.

WDYT:
At a lower conceptual level and in the implementation we can have the notion of *Resource Functions*.  *Resource Functions* are executables that accept an arbitrary number of Resources on STDIN and emit an arbitrary number of Resources on STDOUT.  
TBD: Flags and environment variables are propagated to *Resource Functions*.

At a higher level - *Kustomize* execution may be broken up into 3 distinct phases:

- generation
- transformation
- validation (doesn't exist yet)

Users may extend Kustomize by providing specific types of *Resource Functions* which to implement specific patterns based on the phase they are executed during `generator`, `transformer`, `validator`, etc.

This prevents users from developing a Turing complete pipeline on top of *Resource Functions*, for which they can use other tools specifically targeted at pipelines.

Reply:
sgtm.

> This prevents users from developing a Turing complete pipeline on top of Resource Functions

This is where humanity will make its final stand against the robot invasion :stuck_out_tongue:

Reply:
> This is where humanity will make its final stand against the robot invasion 😛

I have always viewed that as my true calling.

Reply:
curious how would this be/notbe turing complete